### DISCUSSION TITLE
How to extend example with scalar DateTime

### DISCUSSION THREAD

**Discussion by kerkhoffc** [graphql-java/graphql-java#2891] [Q&A] [answered]
The [example](https://www.graphql-java.com/tutorials/getting-started-with-spring-boot/) easily ran on my machine.
I would like to learn how to apply the type scalar DateTime to add e.g. "lastAccessed: DateTime" to the book.
Adding "scalar DateTime" to the schema file didn't solve the exception "There is no scalar implementation for the named  'DateTime' scalar type".

How can I use DateTime in this example?

Best,
Carsten

**Comment by dondonz** [comment, marked as answer]
Hello, as DateTime technically isn't in the GraphQL spec, we've added it in this extended scalar library https://github.com/graphql-java/graphql-java-extended-scalars

More context in this blog post: https://www.graphql-java.com/blog/introducing-extended-scalars

**Reply by kerkhoffc** [reply]
Thank you for the quick response :-)
I checked the links and added the dependency to version 18.1 to my pom.xml.
In the schema file I added a new field `issueDate: DateTime` with and without additional line `scalar DateTime`
The trace during app-start tells me that the field type DateTime is not present.
Sorry for my blindness or wrong question - what is missing to apply DateTime scalar?

org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'routerFunctionMapping' defined in class path resource [org/springframework/boot/autoconfigure/web/servlet/WebMvcAutoConfiguration$EnableWebMvcConfiguration.class]: Invocation of init method failed; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'graphQlRouterFunction' defined in class path resource [org/springframework/boot/autoconfigure/graphql/servlet/GraphQlWebMvcAutoConfiguration.class]: Unsatisfied dependency expressed through method 'graphQlRouterFunction' parameter 0; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'graphQlHttpHandler' defined in class path resource [org/springframework/boot/autoconfigure/graphql/servlet/GraphQlWebMvcAutoConfiguration.class]: Unsatisfied dependency expressed through method 'graphQlHttpHandler' parameter 0; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'webGraphQlHandler' defined in class path resource [org/springframework/boot/autoconfigure/graphql/servlet/GraphQlWebMvcAutoConfiguration.class]: Unsatisfied dependency expressed through method 'webGraphQlHandler' parameter 0; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'executionGraphQlService' defined in class path resource [org/springframework/boot/autoconfigure/graphql/GraphQlAutoConfiguration.class]: Unsatisfied dependency expressed through method 'executionGraphQlService' parameter 0; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'graphQlSource' defined in class path resource [org/springframework/boot/autoconfigure/graphql/GraphQlAutoConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.graphql.execution.GraphQlSource]: Factory method 'graphQlSource' threw exception; nested exception is SchemaProblem{errors=[**The field type 'DateTime' is not present** when resolving type 'Book' [@7:1]]}

**Reply by dondonz** [reply]
There's only one more step - adding the `DateTime` scalar to the `RuntimeWiring`. The error message basically says, you've got a `DateTime` field in the schema but the GraphQL Java engine doesn't know what to do with it. Unfortunately the engine doesn't automatically pick up the `DateTime` scalar from the library, you'll have to register it manually.

With Spring for GraphQL, add config like this
```
@Configuration
public class GraphQlConfig {
    @Bean
    public RuntimeWiringConfigurer runtimeWiringConfigurer() {
        return wiringBuilder -> wiringBuilder
                .scalar(ExtendedScalars.DateTime);
    }
}
```

And it should work. Make sure to include `scalar DateTime` in your schema as well.

More info from the Spring documentation: https://docs.spring.io/spring-graphql/docs/current/reference/html/#execution-graphqlsource-runtimewiring-configurer

**Reply by kerkhoffc** [reply]
Wonderful - it works

**Reply by maurocap** [reply]
it worked here too. Thanks!

**Comment by kerkhoffc** [comment]
I understood that I then have to use datafetchers and all the stuff instead of easy access with `@QueryMapping` ;-)
